### ISSUE TITLE
Question: Passing ID Token to Web Api

### ISSUE THREAD

**Issue by ghost** [rlondner/okta-aspnet-spa_webapi-oidc#1]
Hi
Thanks for this sample.
The SPA is sending the ID Token to the Web API with each request. Is this a good practice? I thought it's the Access Token that should be sent to the Web API?
Thanks
Steve

**Comment by raphaellondner-okta** [comment]
Hi Steve, I agree with you that it's not necessarily a good practice. However, I built this sample when our full OAuth product wasn't ready and I used the ID Token in the meantime. I will soon update this sample to use the access token indeed (which we will also soon expose within the Okta Sign In Widget). Thank you and stay tuned for an update soon!

**Comment by ghost** [comment]
Do you think you'll have time to make it use Access Token soon Raphael? We're really keen to see this example as we can't think of a way for a Web API to validate the Access Token since it is opaque to the app. Even if you could provide some pseudo code that would be immense help to us!

**Comment by raphaellondner-okta** [comment]
Hi Steve, this repo is a bit outdated and you should check out the [latest repo](https://github.com/oktadeveloper/okta-oauth-spa-authjs-osw) on our official OktaDeveloper GitHub org. That sample implements calls to the Web API with the Access Token instead of the ID Token (available only in the Auth JS SDK, not yet in the Sign-In Widget). Note that you'll need access to our API Access Management feature so if it's not available in your org (check the PDF documentation of the repo I pointed to above for details), please contact Okta at beta at okta dot come.

BTW, note that I got some internal feedback after you asked this original question and it turns out the use of ID Token from a SPA calling a Web API is industry-wide accepted and even [recommended by Microsoft](https://azure.microsoft.com/en-us/documentation/articles/active-directory-authentication-scenarios/#single-page-application-spa). I thought I would that information on to you so that you don't stay with the impression that using the ID token for that purpose (in a SPA scenario) is not a good practice.

**Comment by ghost** [comment]
Thanks Raphael for the quick and helpful response. I'll look at that repo from now.

Using ID Token or Access Token. It's interesting that you say ID token usage is industry-wide accepted but now the latest Okta sample repo uses Access Token. I asked it [on stackoverflow and a Microsoft MVP said he wouldn't recommend it](http://stackoverflow.com/questions/38417483/openid-connect-implicit-flow-resource-server#comment64274047_38425166). I can see why the use of the ID Token wouldn't be great, as it means passing user details in every request.

**Comment by raphaellondner-okta** [comment]
Steve, I may have exaggerated my claims on industry-side acceptance a bit here, but that was my impression when I talked about that topic with my colleague. Now, I can't talk for Microsoft, but it's clear from their auth scenarios that they don't seem to have a problem making a call to a web API from a SPA using the ID token. Microsoft MVPs are not Microsoft employees, so I'd suggest you to try to get first-hand feedback from MS' product management team. Note also there's a fine line between acceptance and recommendation ;-)

Last but not least, our new repo actually gives you the ability to configure which type of token to pass to your back-end API. Note also that if you need to call multiple, separate APIs, you would better use access tokens indeed, as you will be able to get multiple access tokens soon (one for each back-end API) and have them carry custom scopes that will depend on a set of authorization rules you'll be able to configure on the Okta side to essentially be able to perform authorization in your back-end API.

In short, use the ID token only in the simplest case where your SPA only needs to call one single API and your API doesn't have the need for custom scopes (and can do away with Okta's "groups" claim for instance to perform API authorization available in the ID token). That arguably only handles limited scenarios, but the options are there for you to choose :-)
